render-ws-spark-client/src/main/java/org/janelia/render/client/spark/zspacing/ZPositionCorrectionClient.java

```diff
@@ -0,0 +1,206 @@
+            @Override
+            public void runClient(final String[] args) throws Exception {
+                final Parameters parameters = new Parameters();
+                parameters.parse(args);
+                final ZPositionCorrectionClient client = new ZPositionCorrectionClient();
+                client.createContextAndRun(parameters);
+            }
+        };
+        clientRunner.run();
+    }
+
+    /** Empty constructor required for alignment pipeline steps. */
+    public ZPositionCorrectionClient() {
+    }
+
+    /** Create a spark context and run the client with the specified parameters. */
+    public void createContextAndRun(final Parameters clientParameters) throws IOException {
+        final SparkConf conf = new SparkConf().setAppName(getClass().getSimpleName());
+        try (final JavaSparkContext sparkContext = new JavaSparkContext(conf)) {
+            LOG.info("run: appId is {}", sparkContext.getConf().getAppId());
+            deriveAndSolveCrossCorrelationData(sparkContext, clientParameters);
+        }
+    }
+
+    /** Validates the specified pipeline parameters are sufficient. */
+    @Override
+    public void validatePipelineParameters(final AlignmentPipelineParameters pipelineParameters)
+            throws IllegalArgumentException {
+        AlignmentPipelineParameters.validateRequiredElementExists("zSpacing",
+                                                                  pipelineParameters.getZSpacing());
+    }
+
+    /** Run the client as part of an alignment pipeline. */
+    @Override
+    public void runPipelineStep(final JavaSparkContext sparkContext,
+                                final AlignmentPipelineParameters pipelineParameters)
+            throws IllegalArgumentException, IOException {
+
+        final Parameters clientParameters = new Parameters();
+        clientParameters.multiProject = pipelineParameters.getMultiProject();
+        clientParameters.zSpacing = pipelineParameters.getZSpacing();
+        deriveAndSolveCrossCorrelationData(sparkContext, clientParameters);
+    }
+
+    private void deriveAndSolveCrossCorrelationData(final JavaSparkContext sparkContext,
+                                                    final Parameters clientParameters)
+            throws IOException {
+
+        final int comparisonRange = clientParameters.zSpacing.getComparisonRange();
+
+        LOG.info("deriveAndSolveCrossCorrelationData: entry, clientParameters={}, comparisonRange={}",
+                 clientParameters, comparisonRange);
+
+        final MultiProjectParameters multiProjectParameters = clientParameters.multiProject;
+
+        final List<StackWithZValues> stackWithAllZValuesList = multiProjectParameters.buildListOfStackWithAllZ();
+
+        final long totalNumberOfZValues = stackWithAllZValuesList.stream()
+                .map(stackWithZ -> (long) stackWithZ.getzValues().size())
+                .reduce(0L, Long::sum);
```

Comment:
Why are we using `long` here? Are we expecting more than 2 billion z-layers? Also, it seems that this is converted to `int` further down, anyway.

Reply:
I was just being super-careful since this supports multi-project / multi-stack runs.  Although it is very unlikely we'll ever need to process billions of z-layers, it didn't seem to hurt to use long here so that we don't have to worry about it.  The long is later converted to an int after batching (so after being divided by cluster parallelism) and will properly throw an exception if that conversion produces an overflow.